README.md
# ddot_rest_server

HISTORY.rst
=======
History
=======

3.2.0 (2019-07-13)
------------------

* Added new parameter **networkattributes** for POST endpoint that
  lets caller set network attributes for base network uploaded to NDEx

* Added new endpoint /<task id>/rawclusteringoutput that contains
  raw output from clustering algorithm used


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for ddot_rest_server."""


tests/test_ddot_restserver.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `ddot_rest_server` package."""

import os
import json
import unittest
import shutil
import tempfile
import io
import uuid
import re
from werkzeug.datastructures import FileStorage
import ddot_rest_server
from ddot_rest_server import ErrorResponse


class TestDdot_rest(unittest.TestCase):
    """Tests for `ddot_rest_server` package."""

    def setUp(self):
        """Set up test fixtures, if any."""
        self._temp_dir = tempfile.mkdtemp()
        ddot_rest_server.app.testing = True
        ddot_rest_server.app.config[ddot_rest_server.JOB_PATH_KEY] = self._temp_dir
        ddot_rest_server.app.config[ddot_rest_server.WAIT_COUNT_KEY] = 1
        ddot_rest_server.app.config[ddot_rest_server.SLEEP_TIME_KEY] = 0
        self._app = ddot_rest_server.app.test_client()

    def tearDown(self):
        """Tear down test fixtures, if any."""
        shutil.rmtree(self._temp_dir)

    def test_error_response(self):
        er = ErrorResponse()
        self.assertEqual(er.errorCode, '')
        self.assertEqual(er.message, '')
        self.assertEqual(er.description, '')
        self.assertEqual(er.stackTrace, '')
        self.assertEqual(er.threadId, '')
        self.assertTrue(er.timeStamp is not None)

    def test_get_submit_dir(self):
        spath = os.path.join(self._temp_dir, ddot_rest_server.SUBMITTED_STATUS)
        self.assertEqual(ddot_rest_server.get_submit_dir(), spath)

    def test_get_processing_dir(self):
        spath = os.path.join(self._temp_dir, ddot_rest_server.PROCESSING_STATUS)
        self.assertEqual(ddot_rest_server.get_processing_dir(), spath)

    def test_get_done_dir(self):
        spath = os.path.join(self._temp_dir, ddot_rest_server.DONE_STATUS)
        self.assertEqual(ddot_rest_server.get_done_dir(), spath)

    def test_create_task_success(self):
        pdict = {}
        pdict['remoteip'] = '1.2.3.4'
        pdict[ddot_rest_server.ALPHA_PARAM] = 0.01
        pdict[ddot_rest_server.BETA_PARAM] = 0.5
        intfile = FileStorage(stream=io.BytesIO(b'hi there'),
                              filename='yo.txt')
        pdict[ddot_rest_server.INTERACTION_FILE_PARAM] = intfile
        res = ddot_rest_server.create_task(pdict)
        self.assertTrue(res is not None)

        snp_path = os.path.join(ddot_rest_server.get_submit_dir(),
                                pdict['remoteip'], res,
                                ddot_rest_server.INTERACTION_FILE_PARAM)
        self.assertTrue(os.path.isfile(snp_path))

    def test_create_task_submitdir_is_a_file(self):
        open(ddot_rest_server.get_submit_dir(), 'a').close()
        pdict = {}
        pdict['remoteip'] = '1.2.3.4'
        pdict[ddot_rest_server.ALPHA_PARAM] = 0.01
        pdict[ddot_rest_server.BETA_PARAM] = 0.5
        intfile = FileStorage(stream=io.BytesIO(b'hi there'),
                              filename='yo.txt')
        pdict[ddot_rest_server.INTERACTION_FILE_PARAM] = intfile
        try:
            ddot_rest_server.create_task(pdict)
            self.fail('Expected NotADirectoryError')
        except NotADirectoryError:
            pass

    def test_get_task_basedir_none(self):
        self.assertEqual(ddot_rest_server.get_task('foo'), None)
        
    def test_get_task_basedir_not_a_directory(self):
        somefile = os.path.join(self._temp_dir, 'hi')
        open(somefile, 'a').close()
        self.assertEqual(ddot_rest_server.get_task('foo', basedir=somefile), None)

    def test_get_task_for_none_uuid(self):
        self.assertEqual(ddot_rest_server.get_task(None,
                                              basedir=self._temp_dir), None)

    def test_get_task_for_nonexistantuuid(self):
        self.assertEqual(ddot_rest_server.get_task(str(uuid.uuid4()),
                                              basedir=self._temp_dir), None)

    def test_get_task_for_validuuid(self):
        somefile = os.path.join(self._temp_dir, '1')
        open(somefile, 'a').close()
        theuuid_dir = os.path.join(self._temp_dir, '1.2.3.4', '1234')
        os.makedirs(theuuid_dir, mode=0o755)

        someipfile = os.path.join(self._temp_dir, '1.2.3.4', '1')
        open(someipfile, 'a').close()
        self.assertEqual(ddot_rest_server.get_task('1234',
                                              basedir=self._temp_dir),
                         theuuid_dir)

    def test_wait_for_task_uuid_none(self):
        self.assertEqual(ddot_rest_server.wait_for_task(None), None)

    def test_wait_for_task_uuid_not_found(self):
        self.assertEqual(ddot_rest_server.wait_for_task('foo'), None)

    def test_wait_for_task_uuid_found(self):
        taskdir = os.path.join(self._temp_dir, 'done', '1.2.3.4', 'haha')
        os.makedirs(taskdir, mode=0o755)
        self.assertEqual(ddot_rest_server.wait_for_task('haha'), taskdir)

    def test_baseurl(self):
        """Test something."""
        rv = self._app.get('/')
        self.assertEqual(rv.status_code, 200)
        self.assertTrue('Data-Driven Ontology Toolkit' in str(rv.data))

    def test_delete(self):
        rv = self._app.delete(ddot_rest_server.ONTOLOGY_NS +
                              '/hehex')

        self.assertEqual(rv.status_code, 200)
        hehefile = os.path.join(self._temp_dir,
                                ddot_rest_server.DELETE_REQUESTS,
                                'hehex')
        self.assertTrue(os.path.isfile(hehefile))

        # try with not set path
        rv = self._app.delete(ddot_rest_server.ONTOLOGY_NS)
        self.assertEqual(rv.status_code, 405)

        # try with path greater then 40 characters
        rv = self._app.delete(ddot_rest_server.ONTOLOGY_NS +
                              '/' + 'a' * 41)
        self.assertEqual(rv.status_code, 400)

        # try where we get os error
        xdir = os.path.join(self._temp_dir,
                            ddot_rest_server.DELETE_REQUESTS,
                            'hehe')
        os.makedirs(xdir, mode=0o755)
        rv = self._app.delete(ddot_rest_server.ONTOLOGY_NS +
                              '/hehe')
        self.assertEqual(rv.status_code, 500)
        
    def test_post_missing_required_parameter(self):
        pdict = {}
        pdict[ddot_rest_server.ALPHA_PARAM] = 0.4,
        rv = self._app.post(ddot_rest_server.ONTOLOGY_NS +
                            '/', data=pdict,
                            follow_redirects=True)
        self.assertTrue('interactionfile' in rv.json['errors'])

        self.assertEqual(rv.status_code, 400)

    def test_post_create_task_fails(self):
        open(ddot_rest_server.get_submit_dir(), 'a').close()
        pdict = {}
        pdict[ddot_rest_server.ALPHA_PARAM] = 0.5
        pdict[ddot_rest_server.BETA_PARAM] = 1.0
        pdict[ddot_rest_server.INTERACTION_FILE_PARAM] = (io.BytesIO(b'hi there'),
                                                      'yo.txt')
        rv = self._app.post(ddot_rest_server.ONTOLOGY_NS,
                            data=pdict,
                            follow_redirects=True)
        self.assertEqual(rv.status_code, 500)
        self.assertTrue('Error' in rv.json['message'])

    def test_post_ndex(self):
        pdict = {}
        pdict[ddot_rest_server.ALPHA_PARAM] = 0.5
        pdict[ddot_rest_server.BETA_PARAM] = 1.0
        pdict[ddot_rest_server.INTERACTION_FILE_PARAM] = (io.BytesIO(b'hi there'),
                                                      'yo.txt')
        rv = self._app.post(ddot_rest_server.ONTOLOGY_NS,
                            data=pdict,
                            follow_redirects=True)
        self.assertEqual(rv.status_code, 202)
        res = rv.headers['Location']
        self.assertTrue(res is not None)
        self.assertTrue(ddot_rest_server.ONTOLOGY_NS in res)

        uuidstr = re.sub('^.*/', '', res)
        ddot_rest_server.app.config[ddot_rest_server.JOB_PATH_KEY] = self._temp_dir

        tpath = ddot_rest_server.get_task(uuidstr,
                                     basedir=ddot_rest_server.get_submit_dir())
        self.assertTrue(os.path.isdir(tpath))
        jsonfile = os.path.join(tpath, ddot_rest_server.TASK_JSON)
        ifile = os.path.join(tpath, ddot_rest_server.INTERACTION_FILE_PARAM)
        self.assertTrue(os.path.isfile(ifile))
        self.assertTrue(os.path.isfile(jsonfile))
        with open(jsonfile, 'r') as f:
            jdata = json.load(f)

        self.assertEqual(jdata['tasktype'], 'ddot_ontology')
        self.assertEqual(jdata[ddot_rest_server.ALPHA_PARAM], 0.5)
        self.assertEqual(jdata[ddot_rest_server.BETA_PARAM], 1.0)

    def test_get_status_no_submidir(self):
        rv = self._app.get(ddot_rest_server.ONTOLOGY_NS + '/status')
        data = json.loads(rv.data)
        self.assertEqual(data['status'], 'ok')
        self.assertEqual(data['restVersion'],
                         ddot_rest_server.__version__)
        self.assertEqual(len(data['load']), 3)
        self.assertTrue(data['pcDiskFull'], -1)
        self.assertEqual(rv.status_code, 200)

    def test_get_status(self):
        submitdir = ddot_rest_server.get_submit_dir()
        os.makedirs(submitdir, mode=0o755)
        rv = self._app.get(ddot_rest_server.ONTOLOGY_NS + '/status')
        data = json.loads(rv.data)
        self.assertEqual(data['status'], 'ok')
        self.assertEqual(data['restVersion'],
                         ddot_rest_server.__version__)
        self.assertEqual(len(data['load']), 3)
        self.assertTrue(data['pcDiskFull'] is not None)
        self.assertEqual(rv.status_code, 200)
        
    def test_get_id_none(self):
        rv = self._app.get(ddot_rest_server.ONTOLOGY_NS)
        self.assertEqual(rv.status_code, 405)

    def test_get_id_not_found(self):
        done_dir = os.path.join(self._temp_dir,
                                ddot_rest_server.DONE_STATUS)
        os.makedirs(done_dir, mode=0o755)
        rv = self._app.get(ddot_rest_server.ONTOLOGY_NS +
                           '/1234')
        data = json.loads(rv.data)
        self.assertEqual(data[ddot_rest_server.STATUS_RESULT_KEY],
                         ddot_rest_server.NOTFOUND_STATUS)
        self.assertEqual(rv.status_code, 410)

    def test_get_id_found_in_submitted_status(self):
        task_dir = os.path.join(self._temp_dir,
                                ddot_rest_server.SUBMITTED_STATUS,
                                '45.67.54.33', 'qazxsw')
        os.makedirs(task_dir, mode=0o755)
        rv = self._app.get(ddot_rest_server.ONTOLOGY_NS +
                           '/qazxsw')
        data = json.loads(rv.data)
        self.assertEqual(data[ddot_rest_server.STATUS_RESULT_KEY],
                         ddot_rest_server.SUBMITTED_STATUS)
        self.assertEqual(rv.status_code, 200)

    def test_get_id_found_in_processing_status(self):
        task_dir = os.path.join(self._temp_dir,
                                ddot_rest_server.PROCESSING_STATUS,
                                '45.67.54.33', 'qazxsw')
        os.makedirs(task_dir, mode=0o755)
        rv = self._app.get(ddot_rest_server.ONTOLOGY_NS +
                           '/qazxsw')
        data = json.loads(rv.data)
        self.assertEqual(data[ddot_rest_server.STATUS_RESULT_KEY],
                         ddot_rest_server.PROCESSING_STATUS)
        self.assertEqual(rv.status_code, 200)

    def test_get_id_found_in_done_status_no_result_file(self):
        task_dir = os.path.join(self._temp_dir,
                                ddot_rest_server.DONE_STATUS,
                                '45.67.54.33', 'qazxsw')
        os.makedirs(task_dir, mode=0o755)
        rv = self._app.get(ddot_rest_server.ONTOLOGY_NS +
                           '/qazxsw')
        data = json.loads(rv.data)
        self.assertEqual(data['message'],
                         'No result found')
        self.assertEqual(rv.status_code, 500)

    def test_get_id_found_in_done_status_with_result_file_no_task_file(self):
        task_dir = os.path.join(self._temp_dir,
                                ddot_rest_server.DONE_STATUS,
                                '45.67.54.33', 'qazxsw')
        os.makedirs(task_dir, mode=0o755)
        resfile = os.path.join(task_dir, ddot_rest_server.RESULT)
        with open(resfile, 'w') as f:
            f.write('{ "hello": "there"}')
            f.flush()

        rv = self._app.get(ddot_rest_server.ONTOLOGY_NS +
                           '/qazxsw')
        data = json.loads(rv.data)
        self.assertEqual(data[ddot_rest_server.STATUS_RESULT_KEY],
                         ddot_rest_server.DONE_STATUS)
        self.assertEqual(data[ddot_rest_server.RESULT_KEY]['hello'], 'there')
        self.assertEqual(rv.status_code, 200)

    def test_get_id_found_in_done_status_with_result_file_with_task_file(self):
        task_dir = os.path.join(self._temp_dir,
                                ddot_rest_server.DONE_STATUS,
                                '45.67.54.33', 'qazxsw')
        os.makedirs(task_dir, mode=0o755)
        resfile = os.path.join(task_dir, ddot_rest_server.RESULT)
        with open(resfile, 'w') as f:
            f.write('{ "hello": "there"}')
            f.flush()
        tfile = os.path.join(task_dir, ddot_rest_server.TASK_JSON)
        with open(tfile, 'w') as f:
            f.write('{"task": "yo"}')
            f.flush()

        rv = self._app.get(ddot_rest_server.ONTOLOGY_NS +
                           '/qazxsw')
        data = json.loads(rv.data)
        self.assertEqual(data[ddot_rest_server.STATUS_RESULT_KEY],
                         ddot_rest_server.DONE_STATUS)
        self.assertEqual(data[ddot_rest_server.RESULT_KEY]['hello'], 'there')
        self.assertEqual(rv.status_code, 200)

    def test_log_task_json_file_with_none(self):
        self.assertEqual(ddot_rest_server.log_task_json_file(None), None)


ddot_rest_server/__init__.py
# -*- coding: utf-8 -*-

"""Top-level package for ddot_rest_server."""

__author__ = """Chris Churas"""
__email__ = 'churas.camera@gmail.com'
__version__ = '3.2.0'

from datetime import datetime
import os
import uuid
import json
import shutil
import time
import copy
import flask

from flask import Flask, jsonify, request
from flask_restplus import reqparse, Api, Resource, fields, marshal
from flask_limiter import Limiter
from flask_limiter.util import get_remote_address


desc = """The Data-Driven Ontology Toolkit (DDOT) REST Service

The Data-Driven Ontology Toolkit (DDOT) facilitates the inference, analysis, and visualization of biological hierarchies using a data structure called an ontology.


 **NOTE:** This service is experimental. The interface is subject to change.


""" # noqa


DDOT_REST_SETTINGS_ENV = 'DDOT_REST_SETTINGS'
# global api object
app = Flask(__name__)

JOB_PATH_KEY = 'JOB_PATH'
WAIT_COUNT_KEY = 'WAIT_COUNT'
SLEEP_TIME_KEY = 'SLEEP_TIME'
DEFAULT_RATE_LIMIT_KEY = 'DEFAULT_RATE_LIMIT'

app.config[JOB_PATH_KEY] = '/tmp'
app.config[WAIT_COUNT_KEY] = 60
app.config[SLEEP_TIME_KEY] = 10
app.config[DEFAULT_RATE_LIMIT_KEY] = '360 per hour'

app.config.from_envvar(DDOT_REST_SETTINGS_ENV, silent=True)
app.logger.info('Job Path dir: ' + app.config[JOB_PATH_KEY])
ONTOLOGY_NS = 'ontology'

TASK_JSON = 'task.json'
LOCATION = 'Location'
TMP_RESULT = 'result.tmp'
RESULT = 'result.json'
CLUSTEROUT = 'rawcluster.output'

ERROR_PARAM = 'error'
REMOTEIP_PARAM = 'remoteip'


STATUS_RESULT_KEY = 'status'
NOTFOUND_STATUS = 'notfound'
UNKNOWN_STATUS = 'unknown'
SUBMITTED_STATUS = 'submitted'
PROCESSING_STATUS = 'processing'
DONE_STATUS = 'done'
ERROR_STATUS = 'error'

# directory where token files named after tasks to delete
# are stored
DELETE_REQUESTS = 'delete_requests'

# key in result dictionary denoting the
# result data
RESULT_KEY = 'result'
NDEXURL_KEY = 'ndexurl'
HIVIEWURL_KEY = 'hiviewurl'

# key in result dictionary denoting input parameters
PARAMETERS_KEY = 'parameters'
INTERACTION_FILE_PARAM = 'interactionfile'
ALPHA_PARAM = 'alpha'
BETA_PARAM = 'beta'
NDEXSERVER_PARAM = 'ndexserver'
NDEXUSER_PARAM = 'ndexuser'
NDEXPASS_PARAM = 'ndexpass'
NDEXNAME_PARAM = 'ndexname'
HIVIEWURL_PARAM = 'hiviewurl'
NETATTRIB_PARAM = 'networkattributes'


api = Api(app, version=str(__version__),
          title='Data-Driven Ontology Toolkit (DDOT) REST Service',
          description=desc, example='put example here')

# enable rate limiting
limiter = Limiter(
    app,
    key_func=get_remote_address,
    default_limits=[app.config[DEFAULT_RATE_LIMIT_KEY]],
    headers_enabled=True
)

# add rate limiting logger to the regular app logger
for handler in app.logger.handlers:
    limiter.logger.addHandler(handler)

# need to clear out the default namespace
api.namespaces.clear()

ns = api.namespace(ONTOLOGY_NS,
                   description='DDOT REST Service')

app.config.SWAGGER_UI_DOC_EXPANSION = 'list'


def get_uuid():
    """
    Generates UUID and returns as string. With one caveat,
    if app.config[USE_SEQUENTIAL_UUID] is set and True
    then uuid_counter is returned and incremented
    :return: uuid as string
    """
    return str(uuid.uuid4())


def get_submit_dir():
    """
    Gets base directory where submitted jobs will be placed
    :return:
    """
    return os.path.join(app.config[JOB_PATH_KEY], SUBMITTED_STATUS)


def get_processing_dir():
    """
        Gets base directory where processing jobs will be placed
    :return:
    """
    return os.path.join(app.config[JOB_PATH_KEY], PROCESSING_STATUS)


def get_done_dir():
    """
        Gets base directory where completed jobs will be placed

    :return:
    """
    return os.path.join(app.config[JOB_PATH_KEY], DONE_STATUS)


def get_delete_request_dir():
    """
    Gets base directory where delete request token files will be placed
    :return:
    """
    return os.path.join(app.config[JOB_PATH_KEY], DELETE_REQUESTS)


def create_task(params):
    """
    Creates a task by consuming data from request_obj passed in
    and persisting that information to the filesystem under
    JOB_PATH/SUBMIT_DIR/<IP ADDRESS>/UUID with various parameters
    stored in TASK_JSON file and if the 'network' file is set
    that data is dumped to NETWORK_DATA file within the directory
    :param request_obj:
    :return: string that is a uuid which denotes directory name
    """
    params['uuid'] = get_uuid()
    params['tasktype'] = 'ddot_ontology'
    taskpath = os.path.join(get_submit_dir(), str(params[REMOTEIP_PARAM]),
                            str(params['uuid']))
    try:
        original_umask = os.umask(0)
        os.makedirs(taskpath, mode=0o775)
    finally:
        os.umask(original_umask)

    app.logger.debug('interaction file param: ' +
                     str(params[INTERACTION_FILE_PARAM]))
    interfile_path = os.path.join(taskpath, INTERACTION_FILE_PARAM)
    with open(interfile_path, 'wb') as f:
        shutil.copyfileobj(params[INTERACTION_FILE_PARAM].stream, f)
        f.flush()
    os.chmod(interfile_path, mode=0o775)

    params[INTERACTION_FILE_PARAM] = INTERACTION_FILE_PARAM
    app.logger.debug(interfile_path + ' saved and it is ' +
                     str(os.path.getsize(interfile_path)) + ' bytes')

    tmp_task_json = TASK_JSON + '.tmp'
    taskfilename = os.path.join(taskpath, tmp_task_json)
    with open(taskfilename, 'w') as f:
        json.dump(params, f)
        f.flush()
    os.chmod(taskfilename, mode=0o775)
    shutil.move(taskfilename, os.path.join(taskpath, TASK_JSON))
    return params['uuid']


def log_task_json_file(taskpath):
    """
    Writes information about task to logger
    :param taskpath: path to task
    :return: None
    """
    if taskpath is None:
        return None

    tmp_task_json = TASK_JSON
    taskfilename = os.path.join(taskpath, tmp_task_json)

    if not os.path.isfile(taskfilename):
        return None

    with open(taskfilename, 'r') as f:
        data = json.load(f)
        app.logger.info('Json file of task: ' + str(data))


def get_task(uuidstr, iphintlist=None, basedir=None):
    """
    Gets task under under basedir.
    :param uuidstr: uuid string for task
    :param iphintlist: list of ip addresses as strings to speed up search.
                       if set then each
                       '/<basedir>//<iphintlist entry>/<uuidstr>'
                       is first checked and if the path is a directory
                       it is returned
    :param basedir:  base directory as string ie /foo
    :return: full path to task or None if not found
    """
    if uuidstr is None:
        app.logger.warning('Path passed in is None')
        return None

    if basedir is None:
        app.logger.error('basedir is None')
        return None

    if not os.path.isdir(basedir):
        app.logger.error(basedir + ' is not a directory')
        return None

    # Todo: Add logic to leverage iphintlist
    # Todo: Add a retry if not found with small delay in case of dir is moving
    for entry in os.listdir(basedir):
        ip_path = os.path.join(basedir, entry)
        if not os.path.isdir(ip_path):
            continue
        for subentry in os.listdir(ip_path):
            if uuidstr != subentry:
                continue
            taskpath = os.path.join(ip_path, subentry)

            if os.path.isdir(taskpath):
                return taskpath
    return None


def wait_for_task(uuidstr, hintlist=None):
    """
    Waits for task to appear in done directory
    :param uuidstr: uuid of task
    :param hintlist: list of ip addresses to search under
    :return: string containing full path to task or None if not found
    """
    if uuidstr is None:
        app.logger.error('uuid is None')
        return None

    counter = 0
    taskpath = None
    done_dir = get_done_dir()
    while counter < app.config[WAIT_COUNT_KEY]:
        taskpath = get_task(uuidstr, iphintlist=hintlist,
                            basedir=done_dir)
        if taskpath is not None:
            break
        app.logger.debug('Sleeping while waiting for ' + uuidstr)
        time.sleep(app.config[SLEEP_TIME_KEY])
        counter = counter + 1

    if taskpath is None:
        app.logger.info('Wait time exceeded while looking for: ' + uuidstr)

    return taskpath


ERROR_RESP = api.model('ErrorResponseSchema', {
    'errorCode': fields.String(description='Error code to help identify issue'),
    'message': fields.String(description='Human readable description of error'),
    'description': fields.String(description='More detailed description of error'),
    'stackTrace': fields.String(description='stack trace of error'),
    'threadId': fields.String(description='Id of thread running process'),
    'timeStamp': fields.String(description='UTC Time stamp in YYYY-MM-DDTHH:MM.S')
})

TOO_MANY_REQUESTS = api.model('TooManyRequestsSchema', {
    'message': fields.String(description='Contains detailed message about exceeding request limits')
})

RATE_LIMIT_HEADERS = {
 'x-ratelimit-limit': 'Request rate limit',
 'x-ratelimit-remaining': 'Number of requests remaining',
 'x-ratelimit-reset': 'Request rate limit reset time'
}


class ErrorResponse(object):
    """Error response
    """
    def __init__(self):
        """
        Constructor
        """
        self.errorCode = ''
        self.message = ''
        self.description = ''
        self.stackTrace = ''
        self.threadId = ''
        self.timeStamp = ''

        dt = datetime.utcnow()
        self.timeStamp = dt.strftime('%Y-%m-%dT%H:%M.%s')


@api.doc('Runs query')
@ns.route('/', strict_slashes=False)
class RunOntology(Resource):
    """
    Runs ontology
    """
    POST_HEADERS = copy.deepcopy(RATE_LIMIT_HEADERS)
    POST_HEADERS['Location'] = 'URL containing resource/result generated by this request'

    post_parser = reqparse.RequestParser()

    post_parser.add_argument(INTERACTION_FILE_PARAM, type=reqparse.FileStorage,
                             required=True,
                             help='File with pairwise distances between '
                                  'elements.  File should be 3 tab separated '
                                  'columns with 1 undirected edge per line '
                                  'of the format node1, node2, edgeWeight '
                                  '(similarity between node1 and node2)\n\n'
                                  'Example file:\n\n ```Bash\nARL2BP  DDAH1 '
                                  '  0.5101\nARL2BP  REPS1   0.6277\n```\n',
                             location='files')
    post_parser.add_argument(ALPHA_PARAM, type=float, default=0.05,
                             help='Threshold between clusters',
                             location='form')
    post_parser.add_argument(BETA_PARAM, type=float, default=0.5,
                             help='Merge density for overlapping clusters',
                             location='form')
    post_parser.add_argument(NDEXNAME_PARAM, default='DDOTontology',
                             help='Name to use for network stored in NDEx',
                             location='form')
    post_parser.add_argument(NDEXSERVER_PARAM, default='test.ndexbio.org',
                             help='NDEx server to use',
                             location='form')
    post_parser.add_argument(NDEXUSER_PARAM, default='ddot_anon',
                             help='NDEx username',
                             location='form')
    post_parser.add_argument(NDEXPASS_PARAM, default='ddot_anon',
                             help='NDEx password',
                             location='form')
    post_parser.add_argument(HIVIEWURL_PARAM,
                             default='http://hiview-test.ucsd.edu',
                             help='HiView server',
                             location='form')
    post_parser.add_argument(NETATTRIB_PARAM, type=str,
                             help='Updates network attributes for base '
                                  'network generated by DDOT. '
                                  'This field should be set as a json string '
                                  'in CX format for network attributes '
                                  '[{"n": "attrib name", '
                                  '"v": <attrib value>, "d": "string"}]" '
                                  'WARNING: name, description, and version '
                                  'CANNOT be updated this way and'
                                  ' will be ignored',
                             location='form')

    @api.doc('Runs Ontology')
    @api.response(202, 'The task was successfully submitted to the service. '
                       'Visit the URL'
                       ' specified in **Location** field in HEADERS to '
                       'status and results', headers=POST_HEADERS)
    @api.response(400, 'Bad request, an invalid input was passed in')
    @api.response(429, 'Too many requests', TOO_MANY_REQUESTS, headers=RATE_LIMIT_HEADERS)
    @api.response(500, 'Internal server error', ERROR_RESP, headers=RATE_LIMIT_HEADERS)
    @api.expect(post_parser)
    def post(self):
        """
        Submits request

        The service should upon post return 202 and set location to resource to poll for result. Which will
        Match the URL of GET request below.
        """
        app.logger.debug("Post received")

        try:
            params = RunOntology.post_parser.parse_args(request, strict=True)
            params['remoteip'] = request.remote_addr

            res = create_task(params)

            resp = flask.make_response()
            resp.headers[LOCATION] = ONTOLOGY_NS + '/' + res
            resp.status_code = 202
            return resp
        except OSError as ea:
            app.logger.exception('Error creating task due to Exception ' +
                                 str(ea))
            er = ErrorResponse()
            er.message = 'Error creating task due to Exception'
            er.description = str(ea)
            return marshal(er, ERROR_RESP), 500


@ns.route('/<string:id>', strict_slashes=False)
class GetQueryResult(Resource):
    """More class doc here"""
    @api.response(200, 'Successful response from server')
    @api.response(410, 'Task not found')
    @api.response(429, 'Too many requests', TOO_MANY_REQUESTS)
    @api.response(500, 'Internal server error', ERROR_RESP)
    def get(self, id):
        """
        Gets results
        """
        cleanid = id.strip()

        taskpath = get_task(cleanid, basedir=get_submit_dir())

        if taskpath is not None:
            resp = jsonify({STATUS_RESULT_KEY: SUBMITTED_STATUS,
                            PARAMETERS_KEY: self._get_task_parameters(taskpath)})
            resp.status_code = 200
            return resp

        taskpath = get_task(cleanid, basedir=get_processing_dir())

        if taskpath is not None:
            resp = jsonify({STATUS_RESULT_KEY: PROCESSING_STATUS,
                            PARAMETERS_KEY: self._get_task_parameters(taskpath)})
            resp.status_code = 200
            return resp

        taskpath = get_task(cleanid, basedir=get_done_dir())

        if taskpath is None:
            resp = jsonify({STATUS_RESULT_KEY: NOTFOUND_STATUS,
                            PARAMETERS_KEY: None})
            resp.status_code = 410
            return resp

        result = os.path.join(taskpath, RESULT)
        if not os.path.isfile(result):
            er = ErrorResponse()
            er.message = 'No result found'
            er.description = self._get_task_parameters(taskpath)
            return marshal(er, ERROR_RESP), 500

        log_task_json_file(taskpath)
        app.logger.info('Result file is ' + str(os.path.getsize(result)) +
                        ' bytes')

        with open(result, 'r') as f:
            data = json.load(f)

        return jsonify({STATUS_RESULT_KEY: DONE_STATUS,
                        RESULT_KEY: data,
                        PARAMETERS_KEY: self._get_task_parameters(taskpath)})

    def _get_task_parameters(self, taskpath):
        """
        Gets task parameters from TASK_JSON file as
        a dictionary
        :param taskpath:
        :return: task parameters
        :rtype dict:
        """
        taskparams = None
        try:
            taskjsonfile = os.path.join(taskpath, TASK_JSON)

            if os.path.isfile(taskjsonfile):
                with open(taskjsonfile, 'r') as f:
                    taskparams = json.load(f)
                if 'remoteip' in taskparams:
                    # delete the remote ip
                    del taskparams['remoteip']
        except Exception:
            app.logger.exception('Caught exception getting parameters')
        return taskparams

    @api.doc('Creates request to delete query')
    @api.response(200, 'Delete request successfully received')
    @api.response(400, 'Invalid delete request', ERROR_RESP)
    @api.response(429, 'Too many requests', TOO_MANY_REQUESTS)
    @api.response(500, 'Internal server error', ERROR_RESP)
    def delete(self, id):
        """
        Deletes task associated with {id} passed in
        """
        resp = flask.make_response()
        try:
            req_dir = get_delete_request_dir()
            if not os.path.isdir(req_dir):
                app.logger.debug('Creating directory: ' + req_dir)
                try:
                    original_umask = os.umask(0)
                    os.makedirs(req_dir, mode=0o775)
                finally:
                    os.umask(original_umask)
            cleanid = id.strip()
            if len(cleanid) > 40 or len(cleanid) == 0:
                er = ErrorResponse()
                er.message = 'Invalid id'
                er.description = 'id is empty or greater then 40 chars'
                return marshal(er, ERROR_RESP), 400

            with open(os.path.join(req_dir, cleanid), 'w') as f:
                f.write(request.remote_addr)
                f.flush()
            resp.status_code = 200
            return resp
        except Exception as e:
            er = ErrorResponse()
            er.message = 'Caught exception'
            er.description = str(e)
            return marshal(er, ERROR_RESP), 500


@ns.route('/<string:id>/rawclusteringoutput', strict_slashes=False)
class GetQueryResult(Resource):
    """More class doc here"""

    @api.response(200, 'Successful response from server, output will be '
                       'raw output as text')
    @api.response(404, 'Task not found or not complete')
    @api.response(429, 'Too many requests', TOO_MANY_REQUESTS)
    @api.response(500, 'Internal server error', ERROR_RESP)
    def get(self, id):
        """
        If a task has completed returns raw output from clustering
        algorithm with mimetype text/plain.
        """
        cleanid = id.strip()

        taskpath = get_task(cleanid, basedir=get_done_dir())

        if taskpath is None:
            resp = flask.make_response()
            resp.status_code = 404
            return resp

        result = os.path.join(taskpath, CLUSTEROUT)
        if not os.path.isfile(result):
            er = ErrorResponse()
            er.message = 'No output found from clustering algorithm'
            er.description = self._get_task_parameters(taskpath)
            return marshal(er, ERROR_RESP), 500

        return flask.send_file(result, mimetype='text/plain')


class ServerStatus(object):
    """Represents status of server
    """
    def __init__(self):
        """Constructor
        """

        self.status = 'ok'
        self.message = ''
        self.pcDiskFull = 0
        self.load = [0, 0, 0]
        self.restVersion = __version__

        self.pcDiskFull = -1
        try:
            s = os.statvfs(get_submit_dir())
            self.pcDiskFull = int(float(s.f_blocks - s.f_bavail) /
                                  float(s.f_blocks) * 100)
        except Exception:
            app.logger.exception('Caught exception checking disk space')
            self.pcDiskFull = -1

        if self.pcDiskFull >= 90:
            self.status = 'error'
            self.message = 'Disk is full'
        else:
            self.status = 'ok'
        loadavg = os.getloadavg()

        self.load[0] = loadavg[0]
        self.load[1] = loadavg[1]
        self.load[2] = loadavg[2]


@ns.route('/status', strict_slashes=False, doc=False)
class SystemStatus(Resource):
    """
    System status
    """
    statusobj = api.model('StatusSchema', {
        'status': fields.String(description='ok|error'),
        'pcDiskFull': fields.Integer(description='How full disk is in %'),
        'load': fields.List(fields.Float(description='server load'),
                            description='List of 3 floats containing 1 minute,'
                                        ' 5 minute, 15minute load'),
        'restVersion': fields.String(description='Version of REST service')
    })
    @api.doc('Gets status')
    @api.response(200, 'Success', statusobj)
    @api.response(429, 'Too many requests', TOO_MANY_REQUESTS)
    @api.response(500, 'Internal server error', ERROR_RESP)
    def get(self):
        """
        Gets status of service

        """
        ss = ServerStatus()
        return marshal(ss, SystemStatus.statusobj), 200
